### DISCUSSION TITLE
How to fold the XML styles? I would like to get an outline tree view much like in `org-mode`

### DISCUSSION THREAD

**Discussion by kjambunathan** [kjambunathan/org-mode-ox-odt#147] [Q&A]
### How to fold the XML styles? I would like to get an outline tree view much like in `org-mode`

CC @QiangF, please add your suggestion here.  (I am not sure why you removed your snippet)

**Comment by kjambunathan** [comment]
`nxml-mode` already has a bunch of commands related to outlines. These are part of file `nxml-outln.el`. 

```

`nxml-mode` Major Mode Bindings Starting With C-c C-o:

Key             Binding

C-c C-o C-a	`nxml-show-all`
C-c C-o C-c	`nxml-hide-direct-text-content`
C-c C-o C-d	`nxml-hide-subheadings`
C-c C-o C-e	`nxml-show-direct-text-content`
C-c C-o TAB	`nxml-show-direct-subheadings`
C-c C-o C-k	`nxml-show-subheadings`
C-c C-o C-l	`nxml-hide-text-content`
C-c C-o C-o	`nxml-hide-other`
C-c C-o C-r	`nxml-refresh-outline`
C-c C-o C-s	`nxml-show`
C-c C-o C-t	`nxml-hide-all-text-content`
```

For outlines to work the following variables have to be customized 

```
`nxml-section-element-name-regexp` is a variable defined in `nxml-outln.el`.

Its value is
"article\\|\\(sub\\)*section\\|chapter\\|div\\|appendix\\|part\\|preface\\|reference\\|simplesect\\|bibliography\\|bibliodiv\\|glossary\\|glossdiv"

Regular expression matching the name of elements used as sections.
An XML element is treated as a section if:

- its local name (that is, the name without the prefix) matches
this regexp;

- either its first child element or a descendant of that first child
element has a local name matching the variable
`nxml-heading-element-name-regexp`; and

- its start-tag occurs at the beginning of a line (possibly indented).

  You can customize this variable.

[back]
```

```
`nxml-heading-element-name-regexp` is a variable defined in `nxml-outln.el`.

Its value is "title\\|head\\|style:style"
Original value was "title\\|head"

Regular expression matching the name of elements used as headings.
An XML element is only recognized as a heading if it occurs as or
within the first child of an element that is recognized as a section.
See the variable `nxml-section-element-name-regexp` for more details.

  You can customize this variable.

[back]
```

The problem is the default settings of the above variables are NOT tuned to OpenDocument format ...

**Comment by QiangF** [comment]
The deletion is unintended, I might clicked the wrong button :-(
Outline minor mode is really fast. If anyone needs to add some manual structure to a document, you can modify outline-regexp as follows:

    (defun my-heading()
        "custom heading for all progn mode, heading must ends in
        outline-heading-end-regexp,which is : for python mode
        * in heading to differentiate from comment, custom heading has higher level "
        (when (and (not (equal major-mode 'org-mode)) comment-start)
        (make-local-variable 'my-outline-regexp)
        (make-local-variable 'my-outline-max-level)
        (setq my-outline-max-level 8)
        (let ((my-heading-start comment-start))
            (setq my-outline-regexp (concat (regexp-quote my-heading-start) "[*]\\{1,8\\}"))
            (setq outline-heading-alist (list)) ;; make outline promote work
            (let ((level 0)
                (level-prefix my-heading-start))
            (while (< level my-outline-max-level)
                (setq outline-heading-alist (cons (cons level-prefix level) outline-heading-alist)
                    level (1+ level)
                    level-prefix (concat level-prefix "*")))
            (setq outline-heading-alist (nreverse outline-heading-alist)))
            (setq outline-regexp (concat my-outline-regexp "\\|" outline-regexp)
                outline-level (lambda()
                                (let* ((data (match-data))
                                        (start (car data))
                                        (end (cadr data))
                                        (level (- end start)))
                                    (if (looking-at my-outline-regexp)
                                        (- level 2)                   ;; subtract two spaces
                                        (+ level my-outline-max-level)) ;; 8 is the maximum custom heading levels
                                    ))))))
    (add-hook 'outline-minor-mode-hook 'my-heading)

Maybe it's a nice idea to add indented heading support.
